Sources/MySQLKit/MySQLDataDecoder.swift

```diff
@@ -34,81 +29,60 @@ public struct MySQLDataDecoder {
     public func decode<T>(_ type: T.Type, from data: MySQLData) throws -> T
         where T: Decodable
     {
+        // If `T` can be converted directly, just do so.
         if let convertible = T.self as? MySQLDataConvertible.Type {
             guard let value = convertible.init(mysqlData: data) else {
-                throw DecodingError.typeMismatch(T.self, DecodingError.Context(
+                throw DecodingError.typeMismatch(T.self, .init(
                     codingPath: [],
-                    debugDescription: "Could not convert MySQL data to \(T.self): \(data)",
-                    underlyingError: nil
+                    debugDescription: "Could not convert MySQL data to \(T.self): \(data)"
                 ))
             }
             return value as! T
         } else {
-            return try T.init(from: _Decoder(data: data, json: self.json))
+            // Probably either a JSON array/object or an enum type not using @Enum. See if it can be "unwrapped" as a
+            // single-value decoding container, since this is much faster than attempting a JSON decode; this will
+            // handle "box" types such as `RawRepresentable` enums and `Optional` (if Optional didn't already directly
+            // conform), but still allow falling back to JSON.
+            do {
+                return try T.init(from: GiftBoxUnwrapDecoder(decoder: self, data: data))
+            } catch DecodingError.dataCorrupted {
+                // Couldn't unwrap it either. Fall back to attempting a JSON decode.
+                return try self.json.decode(T.self, from: data.buffer!)
+            }
         }
     }
     
-    private final class _Decoder: Decoder {
```

Comment:
Love to see this cleaned up, but what's the performance improvement you're seeing?

Reply:
Most of it comes from not going through the `Codable` paths as often, and doing less work in those paths, along with fewer data copies.